### PR TITLE
WIP: Add fixed altitude offboard yaw ctl

### PR DESCRIPTION
**Pull request by yasasa**
## What it does:
Provides a way to control the plane in offboard mode with just yaw rate. Use the tecs to keep the altitude set point and then calculate and publish the roll set point for the given yaw rate.

## How I tested:
I tested using gazebo and mavros. Publishing a position setpoint to the topic `/mavros/setpoint_position/local` with the yaw rate and z coordinate. mavros takes input in ned so make sure the alitutde is negative. You should observe the plane turning, while attempting to reach target altitude.

## Things to do: 
- Using `roll_limit parameter` to constrain the bank angle used to turn but it seems to not be strict enough as with high yaw rates the plane will not sustain altitude.
- Numerical sanity checks on `airspeed` and `yawspeed` maybe? not super necessary since the roll angle is constrained regardless.
- I changed the condition of velocity control to only occur if not in offboard mode, probably need to think of a better way to do this since the offboard mode control flag is used almost nowhere inside the control modules.

### CODE REVIEW

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 1]
```diff
@@ -651,6 +651,7 @@ void FixedwingAttitudeControl::run()
 						control_input.pitch_rate_setpoint = _pitch_ctrl.get_desired_rate();
 						control_input.yaw_rate_setpoint = _yaw_ctrl.get_desired_rate();
 
+
```
Comment (anassinator) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: remove all the unnecessary additions

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1303,10 +1305,47 @@ FixedwingPositionControl::control_position(const math::Vector<2> &curr_pos, cons
...
+					   radians(_parameters.pitch_limit_min) - _parameters.pitchsp_offset_rad,
+					   radians(_parameters.pitch_limit_max) - _parameters.pitchsp_offset_rad,
+					   _parameters.throttle_min,
+					   _parameters.throttle_max,
+					   throttle,
+					   false,
+					   radians(_parameters.pitch_limit_min));
+
+		// Calculate the bank angle needed to achieve the yaw rate
+		float roll_sp = atan2(pos_sp_curr.yawspeed * speed, 9.81f);
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: is there no gravity constant?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -1303,10 +1305,47 @@ FixedwingPositionControl::control_position(const math::Vector<2> &curr_pos, cons
...
+		float throttle = _parameters.throttle_cruise;
+
+		if (PX4_ISFINITE(pos_sp_curr.cruising_speed) &&
+		    pos_sp_curr.cruising_speed > 0.1f) {
+			airspeed = pos_sp_curr.cruising_speed;
+		}
+
+		if (PX4_ISFINITE(pos_sp_curr.cruising_throttle) &&
+		    pos_sp_curr.cruising_throttle > 0.01f) {
+
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: remove

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp [hunk 4]
```diff
@@ -259,6 +259,7 @@ class FixedwingPositionControl final : public control::SuperBlock, public Module
 	enum FW_POSCTRL_MODE {
 		FW_POSCTRL_MODE_AUTO,
 		FW_POSCTRL_MODE_POSITION,
+		FW_POSCTRL_MODE_OFFBOARD_ALTITUDE,
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp]: is there a non-altitude offboard mode?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -1303,10 +1305,47 @@ FixedwingPositionControl::control_position(const math::Vector<2> &curr_pos, cons
...
+		   && pos_sp_curr.valid) {
+		/* Offboard altitude mode, control on yaw rate with fixed altitude */
+		_control_mode_current = FW_POSCTRL_MODE_OFFBOARD_ALTITUDE;
+
+		_hold_alt = _global_pos.alt;
+		float airspeed = _parameters.airspeed_trim;
+		float throttle = _parameters.throttle_cruise;
+
+		if (PX4_ISFINITE(pos_sp_curr.cruising_speed) &&
+		    pos_sp_curr.cruising_speed > 0.1f) {
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: why 0.1?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 6]
```diff
@@ -1303,10 +1305,47 @@ FixedwingPositionControl::control_position(const math::Vector<2> &curr_pos, cons
...
+			airspeed = pos_sp_curr.cruising_speed;
+		}
+
+		if (PX4_ISFINITE(pos_sp_curr.cruising_throttle) &&
+		    pos_sp_curr.cruising_throttle > 0.01f) {
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: why 0.01?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 7]
```diff
@@ -687,6 +687,7 @@ FixedwingPositionControl::control_position(const math::Vector<2> &curr_pos, cons
 	}
 
 	if (_control_mode.flag_control_auto_enabled && pos_sp_curr.valid) {
+
```
Comment (anassinator) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: remove

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp (hunk 4 above)
Comment (yasasa) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp]: this is for the control mode of the fixedwingpositioncontroller, in the commander state machine there is an offboard mode.

**Comment by anassinator** [comment]
So that we don't forget. We should be testing this with:

```
coordinate_frame: 1
typemask: 1531
```

This typemask will ignore everything except `yaw_rate` and `position.z` (i.e. altitude). Also since this is in NED the frame, the altitude should be negative.

**Comment by yasasa** [comment]
Closing this for now!

**Pull request closed without merging**